types/redux-pack/index.d.ts

```diff
@@ -1,52 +1,78 @@
-    failure?: Reducer<S>;
-    success?: Reducer<S>;
-    always?: Reducer<S>;
+// MetaPayload differs from ActionMeta in that it is the object that the reducers
+// receive, instead of what is dispatched
+export type MetaPayload<M> = M & {
+    ['redux-pack/LIFECYCLE']?: LIFECYCLEValues;
+    ['redux-pack/TRANSACTION']?: string;
+};
+
+// Incomplete typing
+export type PackActionPayload<Payload, M> = ReduxAction & {
+    payload: Payload;
+    meta: MetaPayload<M>;
+};
+
+export type handlerReducer<S, A> = (state: S, action: A) => S;
+export interface Handlers<S, TSuccessPayload, TErrorPayload, TStartPayload, TMetaPayload> {
+    start?: handlerReducer<S, PackActionPayload<TStartPayload, TMetaPayload>>;
+    finish?: handlerReducer<S, ReduxAction>;
+    failure?: handlerReducer<S, PackActionPayload<TErrorPayload, TMetaPayload>>;
+    success?: handlerReducer<S, PackActionPayload<TSuccessPayload, TMetaPayload>>;
+    always?: handlerReducer<S, ReduxAction>;
 }
+
 export type GetState<S> = () => S;
-export interface ActionMeta<TState = {}, TSuccessPayload = {}, TErrorPayload = {}, TStartPayload = {}> {
+export interface ActionMeta<TFullState = {}, TSuccessPayload = {}, TErrorPayload = {}, TStartPayload = {}> {
     startPayload?: TStartPayload;
-    onStart?(payload: TStartPayload, getState: GetState<TState>): void;
-    onFinish?(resolved: boolean, getState: GetState<TState>): void;
-    onSuccess?(response: TSuccessPayload, getState: GetState<TState>): void;
-    onFailure?(error: TErrorPayload, getState: GetState<TState>): void;
-    ['redux-pack/LIFECYCLE']?: keyof LIFECYCLEValues;
+    onStart?(payload: TStartPayload, getState: GetState<TFullState>): void;
+    onFinish?(resolved: boolean, getState: GetState<TFullState>): void;
+    onSuccess?(response: TSuccessPayload, getState: GetState<TFullState>): void;
+    onFailure?(error: TErrorPayload, getState: GetState<TFullState>): void;
+    ['redux-pack/LIFECYCLE']?: LIFECYCLEValues;
     ['redux-pack/TRANSACTION']?: string;
 }
-export interface Action<TState = {}, TSuccessPayload = {}, TErrorPayload = {}, TStartPayload = {}> extends ReduxAction {
+
+export interface PackError { error: boolean; payload: any; }
+export interface Action<TFullState = {}, TSuccessPayload = {}, TErrorPayload = PackError, TStartPayload = {}, TMetaPayload = {}> extends ReduxAction {
     promise?: Promise<TSuccessPayload>;
     payload?: TSuccessPayload | TErrorPayload | TStartPayload;
-    meta?: ActionMeta<TState, TSuccessPayload, TErrorPayload, TStartPayload>;
+    meta?: ActionMeta<TFullState, TSuccessPayload, TErrorPayload, TStartPayload> & TMetaPayload;
+    // add optional error key to conform to FSA design: https://github.com/redux-utilities/flux-standard-action
+    // note that users of this middleware (using our types) must conform to FSA shaped actions or code will not compile
+    error?: boolean | null;
 }
-export function handle<TState>(
+
+export interface TFullState { [key: string]: any; }
```

Comment:
`TFullState` was originally a param of the handle generic, but as it was only used once in the def, the TSLint rules forced me to define it like this, which I dont thin should be a problem as it allows for an obj of any shape.